### PR TITLE
Introduce proper PHP classes for interacting with block types

### PR DESCRIPTION
**Pull request by felixarntz**
This is a PR for #1321.

It currently does the following:
* Introduce a very basic `WP_Block_Type` class that follows the models of other core classes for entity types, such as `WP_Post_Type` and `WP_Taxonomy`.
* Introduce a `WP_Block_Type_Registry` class as a container for `WP_Block_Type` objects, including methods to register, unregister and retrieve block types. A static `get_instance()` method is provided for access to the main class instance.
* The functions `register_block_type()` and `unregister_block_type()` are now wrappers for methods of the `WP_Block_Type_Registry` class. The `do_blocks()` function has been adjusted to use that class as well. The previous `$wp_registered_blocks` global no longer exists.
* In the warning messages when registering or unregistering block types, it changes the terminology from "Block" to "Block Type" because that is what the developer is actually dealing with.

Further ideas/observations:
* At the moment there is no mechanism in `WP_Block_Type` or `WP_Block_Type_Registry` to ensure the `render` argument is set. We could either continue to have it optional and introduce a check whether it is set, or alternatively require it and not register a block type without the argument. I think I'm leaning towards leaving it optional because otherwise it should not be part of `$args`, but a required parameter.
* It may be useful to also introduce a `WP_Block` class that represents an actual block parsed from the content. I'd rather discuss this in a separate issue / PR though.
* Instead of using `_doing_it_wrong()` it might be a better idea to return a `WP_Error` object if something goes wrong when registering or unregistering a block type.
* The doc blocks for `parse_block_attributes()` and `do_blocks()` do not follow the WordPress documentation standards. We can fix this here, but it could also be part of another PR if that's preferred.

### CODE REVIEW

**Comment by felixarntz** [comment]
I've not yet adjusted unit tests to the changes. Will do that later this week.

File: lib/class-wp-block-type.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+		/**
+		 * Filters the arguments for registering a block type.
+		 *
+		 * @since 0.2.0
+		 *
+		 * @param array  $args       Array of arguments for registering a block type.
+		 * @param string $block_type Block type key.
+		 */
+		$args = apply_filters( 'register_block_type_args', $args, $this->name );
```
Comment (westonruter) [inline lib/class-wp-block-type.php]: What's the use case for this?

File: lib/class-wp-block-type.php [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+		$defaults = array(
+			'render' => null,
+		);
+
+		$args = array_merge( $defaults, $args );
+
+		$args['name'] = $this->name;
+
+		foreach ( $args as $property_name => $property_value ) {
```
Comment (westonruter) [inline lib/class-wp-block-type.php]: This will set props even if they are undeclared. Would that be desired? Should something like this be done instead: https://github.com/WordPress/wordpress-develop/blob/4.8.0/src/wp-includes/class-wp-customize-setting.php#L179-L184

File: lib/class-wp-block-type-registry.php [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 0.2.0
+	 * @access private
+	 * @static
+	 * @var WP_Block_Type_Registry|null
+	 */
+	private static $instance = null;
```
Comment (westonruter) [inline lib/class-wp-block-type-registry.php]: Is this going to facilitate testing since the instance can't be cleared?

File: lib/class-wp-block-type.php (hunk 1 above)
Comment (felixarntz) [inline lib/class-wp-block-type.php]: I liked the idea of opening up block types for extensions. A similar filter is also available when registering post types, and it could be used here to add custom block type arguments that could be pre-populated by other default arguments.

If you think we should leave this out for now, I'm not opposed. It's probably too early to figure out a specific use-case where this can come in handy.

File: lib/class-wp-block-type.php (hunk 2 above)
Comment (felixarntz) [inline lib/class-wp-block-type.php]: I think it should, in order to allow setting properties that are unsupported by core, but possibly are by plugins doing something custom with Gutenberg. `WP_Post_Type` and `WP_Taxonomy` work the same way, and it has proven to make sense there.

File: lib/class-wp-block-type-registry.php (hunk 3 above)
Comment (felixarntz) [inline lib/class-wp-block-type-registry.php]: I think testing should be possible as we could manually spin up instances of the `WP_Block_Type_Registry` class for most test cases, as we wouldn't need to use the main instance. Once tests have been added, we'll know for sure.

File: lib/class-wp-block-type.php (hunk 1 above)
Comment (westonruter) [inline lib/class-wp-block-type.php]: I think it would be better to only add the filter when we know we need it, rather than add filters up front when we may not know that they will ever be used.

File: lib/class-wp-block-type.php (hunk 2 above)
Comment (westonruter) [inline lib/class-wp-block-type.php]: In that case would it not be better to allow `WP_Block_Type` to be extensible so that the subclasses could be passed to the `register` method as opposed to just the args? This would be similar to how `WP_Customize_Manager::add_control()` allows you to pass the ID and args, or just a previously instantiated `WP_Customize_Control`.

Comment (felixarntz) [inline lib/class-wp-block-type.php]: That might be a better idea, but I'm not entirely sure. Allowing all `$args` to be set as property especially makes sense with a filter such as `register_block_type_args` (which per the above we're gonna remove for now, but might re-add it again at some point).

I like the idea of passing a custom instance directly (to support sub-classes), but supporting any `$args` key to be set as property makes it possible (in collaboration with a filter) to provide extension points that could affect _all_ block types, regardless of their class. Let's wait for more opinions on this, I would say.

**Comment by BE-Webdesign** [comment]
Hi @felixarntz,

Great work so far! I like that we are moving towards a more class based approach, and it seems to be moving quick! A quick suggestion: it could be an interesting approach to change the `WP_Block_Type_Registry->register()` signature from `( $name, $args )` to just `( $block_type )`, then the registry could act as more of a collection of the blocks that are being registered. This way the registry is not concerned about what a block is and more with whether it has been registered or not. This makes adding a helper method of `WP_Block_Type_Registry->register_block_types()` easy to implement as well. `ArrayAccess` could also potentially be used instead of `WP_Block_Type_Registry->registered_block_types`, the only caveat being that `ArrayAccess` is finicky in PHP. Let me know your thoughts!

**Comment by westonruter** [comment]
When introducing a `WP_Block` it should look the same as blocks parsed in JS, with the attributes, content, and children.

**Comment by felixarntz** [comment]
@BE-Webdesign I generally like the idea of providing a block type in `register_block_type()` (and `WP_Block_Type_Registry::register()`), however it would be a bit confusing to require a `WP_Block_Type` there while requiring a block type name for the unregistering method. Not heavily against this, but it could be confusing to WordPress developers who're used to other core APIs.

Given that we're not used to this behavior of functions, I rather suggest simply supporting a `WP_Block_Type` instance being passed in the registration method (instead of requiring it) - basically `if ( is_a( $name, 'WP_Block_Type' ) ) ...`. This gives us the best of both worlds.

Supporting registration of multiple block types in one might be useful, we could add a simple wrapper for the singular registration method to support this. Not entirely sure if we _need_ to have this though.

@westonruter That makes sense. As said before, I'd prefer if we worked on this in another PR after this one is ready.

File: lib/blocks.php [hunk 4]
```diff
@@ -9,64 +9,34 @@
...
- * Unregisters a block.
+ * Unregisters a block type.
  *
- * @param  string $name Block name.
- * @return array        The previous block value, if it has been
- *                      successfully unregistered; otherwise `null`.
+ * @since 0.1.0
+ *
+ * @param string $name Block type name including namespace.
+ * @return WP_Block_Type|false The unregistered block type on success, or false on failure.
```
Comment (GaryJones) [inline lib/blocks.php]: What benefit is there to returning a `WP_Block_Type` object just after it has been unregistered?
In case someone wants to amend it and re-register it?

File: lib/class-wp-block-type-registry.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	/**
+	 * Retrieves a registered block type.
+	 *
+	 * @since 0.2.0
+	 * @access public
+	 *
+	 * @param string $name Block type name including namespace.
+	 * @return WP_Block_Type|null The unregistered block type, or null if it is not registered.
```
Comment (GaryJones) [inline lib/class-wp-block-type-registry.php]: Typo: `unregistered` => `registered`.

File: lib/blocks.php (hunk 4 above)
Comment (westonruter) [inline lib/blocks.php]: It's a shortcut to avoid having to do a separate getter call, yes. This was also implemented in the Customizer for calls like `WP_Customize_Manager::add_setting()`. Previously it returned `void` but not it returns the `WP_Customize_Setting` instance which is particularly useful if you provided an `$args` array instead of a subclassed instance of `WP_Customize_Setting`. See https://core.trac.wordpress.org/ticket/34596

Comment (westonruter) [inline lib/blocks.php]: Oh, woops. I missed that this is in the _unregister_ call. Nevertheless, seems useful to return. As essentially if `false` is the return value then nothing was unregistered, whereas if it is truthy (an instance of `WP_Block_Type` then the unregistration was successful.

Comment (GaryJones) [inline lib/blocks.php]: I get the value of something truthy being returned to indicate a successful unregister, but why not `true`? i.e. what could a developer a be reasonably expected to do with this object (other than tweak and re-register)? Is there any drawback in just returning `true`? Do other situations in WP return something more than a boolean when something has been unregistered?

Comment (felixarntz) [inline lib/blocks.php]: I think it's a good practice to return the data that has been removed/unregistered, to allow further actions with it, for example tweaking and re-registering or passing it to a custom hook.

It's a bit similar to how several PHP array functions work like `array_shift()`, `array_pop()` or `array_splice()`: The moment you remove something, also return what has been removed on success.

File: lib/class-wp-block-type-registry.php (hunk 5 above)
Comment (felixarntz) [inline lib/class-wp-block-type-registry.php]: Fixed in a40813297a55ce89018ed1ddacdf82939878d375

**Comment by felixarntz** [comment]
In ecc6dbb, I added a new test class specific to methods of `WP_Block_Type_Registry` which we can test unaware of WordPress' global state. Most of the tests in there were previously part of the `Registration_Test` class and have been migrated over to using the new class methods directly. The `Registration_Test` class now only contains tests to ensure that the `register_block_type()` and `unregister_block_type()` functions properly wrap the respective methods of the `WP_Block_Type_Registry` main instance.

In the `Dynamic_Blocks_Render_Test` I just made a few tweaks to adjust to the new structure and added a missing call to `parent::tearDown()` which caused unexpected test behavior.

**Comment by felixarntz** [comment]
What's left here? Should we add support for passing a `WP_Block_Type` to the registration method in this PR, or should it be discussed separately afterwards? And do we want to have more methods and argument validation in `WP_Block_Type`?

File: lib/class-wp-block-type-registry.php [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	 *
+	 * @since 0.2.0
+	 * @access public
+	 *
+	 * @param string $name Block type name including namespace.
+	 * @return WP_Block_Type|null The registered block type, or null if it is not registered.
+	 */
+	public function get_registered( $name ) {
+		if ( ! $this->is_registered( $name ) ) {
+			return null;
```
Comment (GaryJones) [inline lib/class-wp-block-type-registry.php]: Instead of having mixed return types, is there an advantage to returning an instance of a `WP_Null_Block_Type` (not yet defined)?

**Comment by westonruter** [comment]
> Should we add support for passing a `WP_Block_Type` to the registration method in this PR, or should it be discussed separately afterwards? And do we want to have more methods and argument validation in `WP_Block_Type`?

I think passing subclasses of `WP_Block_Type` should be supported up front.

File: lib/class-wp-block-type-registry.php (hunk 6 above)
Comment (felixarntz) [inline lib/class-wp-block-type-registry.php]: That is a good question. Generally returning something like you say (implementing a general block type interface) makes sense, however these practices have never been followed by WordPress core. I'm wary of introducing such a pattern here as it contradicts of what people are used from working with core. I think such a topic belongs into a discussion of a bigger scope, which I'd love to have a working group for at some point (we had a similar discussion at WCEU).

**Comment by felixarntz** [comment]
d45da39 fixes the merge conflicts that have come up with recent Gutenberg upstream changes.

547b89b introduces a `WP_Block_Type::render()` method. This provides a clear separation of concerns and allows overriding the method directly if necessary (for example in a `WP_Block_Type` sub-class). Furthermore the method now handles the case properly that a `render` argument is not provided, which would previously result in a fatal error. It now returns the raw content instead, which also fixes #1631 as it makes that data available to the actual render callback as well. Several unit tests for that functionality as well as for the `set_props()` method have been added as well.

**Comment by felixarntz** [comment]
75a065a adds support for passing a `WP_Block_Type` instance (or sub-class) to `register_block_type()`. When the function is used like that, the second parameter will be ignored. Documentation has been adjusted accordingly, and a unit test has been added.

I think once these changes have been reviewed and accepted, the PR should be good. I'd prefer if we work on further enhancements to the new infrastructure through separate issues / PRs.

File: lib/class-wp-block-type.php [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+	public $name;
+
+	/**
+	 * Block type render callback.
+	 *
+	 * @since 0.4.0
+	 * @access public
+	 * @var callable
+	 */
+	public $render;
```
Comment (westonruter) [inline lib/class-wp-block-type.php]: This should perhaps be called `render_callback` to help disambiguate from `render` the method. This is similar to what `WP_Customize_Control` does with its `active` method and `active_callback` member variable.

See https://github.com/WordPress/wordpress-develop/blob/a16d16449ddee6a3c13a65fd02536a6071c04d70/src/wp-includes/class-wp-customize-control.php#L271-L273

**Comment by felixarntz** [comment]
6b3ec26 renames the `render` property to `render_callback`.

**Comment by westonruter** [comment]
For some more thoughts on how the PHP API should look, see https://twitter.com/Josh412/status/883061703004631041

**Comment by felixarntz** [comment]
@westonruter I don't think it makes sense to suddenly start using such patterns in a feature project, as part of a relatively unrelated PR.
I certainly prefer the approach @Shelob9 suggests, but as much as I don't like this argument this is not the current "WordPress way". I'm all in for discussing this as a general best practice to enforce and require on future code, but that should happen in a broader scope, for example in the new WordPress PHP Slack channel.

In short: Introducing a WordPress container for a random PR in Gutenberg seems wrong to me. Introducing it after a more thorough general discussion, not connected to a specific code change, makes much more sense I think.

**Comment by felixarntz** [comment]
@westonruter What is left to do here to get ready for merge?

**Comment by westonruter** [comment]
Probably a 👍 from @mtias. 

Note that I was suggesting a similar in the JS API, but it was not endorsed: https://github.com/WordPress/gutenberg/issues/362#issuecomment-313857885

**Comment by felixarntz** [comment]
I'm by far not a JS expert, but there are probably reasons for it not to work the same as in PHP. All I know is that in PHP these classes are a much better practice than globals and random arrays. :)

@mtias What are your thoughts? Any objections, suggestions?

**Comment by nylen** [comment]
Let's merge this soon; thank you @felixarntz for staying on top of it.  The existing global variable was written because it was the quickest way to get the code working; this approach is much better.

Regarding differences between JS and PHP APIs:  I think the main benefit of this approach is encapsulation, which ES modules also give us.  PHP and JS (especially ES6) are very different languages, and these differences should imply different coding styles.

It looks like there are a few minor things that need to be updated.  I'll leave review comments.

File: lib/class-wp-block-type-registry.php [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+		return $block_type;
+	}
+
+	/**
+	 * Unregisters a block type.
+	 *
+	 * @since 0.4.0
+	 * @access public
+	 *
+	 * @param string $name Block type name including namespace.
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Like `register` does, should this function also accept a `WP_Block_Type` object?

File: lib/blocks.php [hunk 9]
```diff
@@ -9,64 +9,37 @@
...
 /**
- * Unregisters a block.
+ * Unregisters a block type.
  *
- * @param  string $name Block name.
- * @return array        The previous block value, if it has been
- *                      successfully unregistered; otherwise `null`.
+ * @since 0.1.0
+ *
+ * @param string $name Block type name including namespace.
```
Comment (nylen) [inline lib/blocks.php]: Should this accept a `WP_Block_Type` instance like `register_block_type` does?

File: lib/class-wp-block-type-registry.php [hunk 10]
```diff
@@ -0,0 +1,170 @@
+<?php
+/**
+ * Blocks API: WP_Block_Type_Registry class
+ *
+ * @package gutenberg
+ * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+ */
+
+/**
+ * Core class used for interacting with block types.
+ *
+ * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+ */
+final class WP_Block_Type_Registry {
+	/**
+	 * Registered block types, as `$name => $instance` pairs.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+	 * @access private
+	 * @var array
+	 */
+	private $registered_block_types = array();
+
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+	 * @access private
+	 * @static
+	 * @var WP_Block_Type_Registry|null
+	 */
+	private static $instance = null;
+
+	/**
+	 * Registers a block type.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/blocks.php [hunk 15]
```diff
@@ -9,64 +9,37 @@
...
 /**
- * Registers a block.
+ * Registers a block type.
  *
- * @param  string $name Block name including namespace.
- * @param  array  $settings Block settings.
-
- * @return array            The block, if it has been successfully registered.
+ * @since 0.1.0
+ * @since 0.4.0 Now also accepts a WP_Block_Type instance as first parameter.
```
Comment (nylen) [inline lib/blocks.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+		$unregistered_block_type = $this->registered_block_types[ $name ];
+		unset( $this->registered_block_types[ $name ] );
+
+		return $unregistered_block_type;
+	}
+
+	/**
+	 * Retrieves a registered block type.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+		$this->registered_block_types[ $name ] = $block_type;
+
+		return $block_type;
+	}
+
+	/**
+	 * Unregisters a block type.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+			return null;
+		}
+
+		return $this->registered_block_types[ $name ];
+	}
+
+	/**
+	 * Retrieves all registered block types.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+	 * @return array Associative array of `$block_type_name => $block_type` pairs.
+	 */
+	public function get_all_registered() {
+		return $this->registered_block_types;
+	}
+
+	/**
+	 * Checks if a block type is registered.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type-registry.php [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+	public function is_registered( $name ) {
+		return isset( $this->registered_block_types[ $name ] );
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type-registry.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 21]
```diff
@@ -0,0 +1,96 @@
+<?php
+/**
+ * Blocks API: WP_Block_Type class
+ *
+ * @package gutenberg
+ * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+/**
+ * Core class representing a block type.
+ *
+ * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+ *
+ * @see register_block_type()
+ */
+class WP_Block_Type {
+	/**
+	 * Block type key.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+	 * @access public
+	 * @var string
+	 */
+	public $name;
+
+	/**
+	 * Block type render callback.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+	 * @var callable
+	 */
+	public $render_callback;
+
+	/**
+	 * Constructor.
+	 *
+	 * Will populate object properties from the provided arguments.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+	public function __construct( $block_type, $args = array() ) {
+		$this->name = $block_type;
+
+		$this->set_props( $args );
+	}
+
+	/**
+	 * Renders the block type output for given attributes and content.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

File: lib/class-wp-block-type.php [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+			return $content;
+		}
+
+		return call_user_func( $this->render_callback, $attributes, $content );
+	}
+
+	/**
+	 * Sets block type properties.
+	 *
+	 * @since 0.4.0
```
Comment (nylen) [inline lib/class-wp-block-type.php]: Should be `@since 0.6.0`.

**Review by nylen — CHANGES_REQUESTED** [review]

**Comment by mtias** [comment]
I haven't had a chance to read this one with attention, but I hope to get to it in the next couple days.

**Comment by felixarntz** [comment]
@nylen Will make the version number changes later today!

**Comment by nylen** [comment]
:+1: There was one other change in there too, about letting the `unregister` functions accept a `WP_Block_Type` instance as well.  I thought it might be helpful to have each version number marked so the feedback will disappear when updated, but maybe that was a bit overkill.

**Comment by mtias** [comment]
Thanks for working on this one. With further structure in place, it begins to highlight a drift between editor-client and back-end regarding _which_ blocks are registered or available at any given point in time (which is also not absolute) that may require us thinking at a bit higher level. In a way, _behaviour_ is ok to be different since they are entirely separate paradigms, but _presence_ may require us to look at a more abstracted way of defining a block—say, a `json` file with minimal configuration like block-type name, title, etc, that specifies entry paths for client and back-end files.

That may become more relevant if we want to have access to a reliable list of potentially available blocks on the back-end for something like storing a default block in Settings → Writing, or for constructing PHP templates declaring blocks that are static (meaning they don't have a representation on the back-end). But equally not sure if this is needed in any case, as there are ways to handle it.

In any case, let's merge this once the version number changes are in.

**Comment by felixarntz** [comment]
@nylen I updated the version numbers and also added support passing a `WP_Block_Type` to `unregister_block_type()`.

**Review by nylen — APPROVED** [review]

**Comment by westonruter** [comment]
> a more abstracted way of defining a block—say, a json file with minimal configuration like block-type name, title, etc, that specifies entry paths for client and back-end files.

This could tie in with using JSON schema to describe block attributes: #1905
